Added additional SetValue methods to TextureBatchItem using a Matrix3…

vpenades opened the pull request:
Added additional SetValue methods to TextureBatchItem using a Matrix3x2 transform generalization.

This allows transform concatenations in scenarios like complex graphics animations, skewing, and viewport size virtualization, which are not possible with any of the other "SetValue" methods.

vpenades committed:
Added additional SetValue methods to TextureBatchItem using a Matrix3x2 transform generalization.

ThomasMiz commented:
Hello and thanks for your contribution!

Is this PR done, or will anything else be added to it?

vpenades commented:
Well, if I would be me I would redirect all the other methods to the one using the matrix3x2 , which would remove a lot of boiler plate code... but you stated you preferred it that way for performance reasons, so on my side this PR is completed.

On a side note, everything I do in 2D ends up requiring an extra layer of transformation for one reason or another, so all 2D rendering calls end up using the matrix3x2 one and not the others.

vpenades commented:
Yout're right, it would need to be exposed in some way.

Would you agree to propagate these two method calls to TextureBatcher?

ThomasMiz commented:
I'm not sure if we're correctly understanding each other

This functionality is already present in the `TextureBatcher.DrawRaw` methods. Is there a use case for this functionality that's not already covered by these methods?

vpenades commented:
It is covered as in, filling the vertices directly, what I was proposing was a helper method that takes the matrix and fills the vertices for you.

but if you consider this as redundant, and given the method inside TextureBatchItem  is internal, then I would agree this PR has no meaning and could proceed to close it.

vpenades commented:
> Sorry for the delay on this one, I've been crazy busy

do't worry, I'm on the same boat, I can only visit this in my spare time.


As it happens, Matrix3x2 is the combination of position, scale and rotation, in fact, these matrices are usually called SRT matrices.

For equivalence:

```c#

batcher.Draw(texture, position, source, color, scale, rotation, origin);

// is equivalent to:

Matrix3x2 xform = Matrix3x2.CreateScale(scale)
                * Matrix3x2.CreateRotation(rotation)
                * Matrix3x2.CreateTranslation(position);
                
batcher.Draw(texture, xform, source, color, origin);


```
"but you have to multiply the 3 components to get the matrix, which adds overhead"

Yes, on naive demos and examples, that's the case.

But, What happens on actual production code?

On more complex scenarios, the most convenient primitive to represent a "2D location" is a Matrix3x2, not a SRT tuple.

- GUIs like WPF or MAUI, under the hood, work with some kind of matrices representation, so you can apply transformations to whole regions, so you can do things like zoomed and scrolled canvas drawing.

- 2D Animated rigs, like characters and so on (think as in Flash animations) also represent character parts as matrices, because each "bone" is a matrix concatenated by its parent bone, using a plain matrix-matrix multiplication.

- Also, on videogames, for example, the game objects are most probably controlled by some kind of physics engine, which happens to work with matrices, so when you want to render the scene, the physics engines typically tell you where every object is located using a matrix, not a SRT tuple.

- Matrices support stretching in cases where you want to fit something on a rectangle that does not have a 1:1 aspect ratio.

- And in general, matrices allow for matrix concatenation which is extremely powerful in graphics programming.

So, in those cases where you have a Matrix and not a SRT tuple in the first place, and the underlaying rendering engine _does not have_ a rendering method that accepts a matrix, what you end up doing is _decomposing the matrix_ into the S, R, T components, which happens to be even more computationally expensive (SQRTs involved) than multiplying matrices.

ThomasMiz commented:
Okay, sounds cool! My notes on this, mostly from curiosity, are:

> = Matrix3x2.CreateScale(scale) * Matrix3x2.CreateRotation(rotation) * Matrix3x2.CreateTranslation(position);

Wait what, what does it mean to multiply two 3x2 matrices, that's not math
After looking at [the struct's code](https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/System.Numerics/System/Numerics/Matrix3x2.cs#L587) it seems like the left matrix gets extended with an [0; 0; 1] columnar vector, and then you can multiply a 3x3 matrix by a 3x2 one, cool

> GUIs like WPF or MAUI, under the hood, work with some kind of matrices representation, so you can apply transformations to whole regions, so you can do things like zoomed and scrolled canvas drawing.

This should be done with a matrix in the vertex shader though, not at sprite-level

> Matrices support stretching in cases where you want to fit something on a rectangle that does not have a 1:1 aspect ratio.

Current `Draw` methods support specifying scale as a `Vector2`, although the `Matrix3x2` one would support skews

> ...the physics engines typically tell you where every object is located using a matrix, not a SRT tuple.

If you have any examples of such engines I'd like to see, as the engines I've used in the past don't give you a matrix

The idea of 2D animated rigs with concatenation sold me though! I do want to see with this PR a small demo project (it can be an additional project in [TrippyTests](https://github.com/SilkCommunity/TrippyGL/tree/master/TrippyTests)) that showcases the new `Matrix3x2` methods by drawing some things with them, including something that uses a rig (something simple is enough, like a bunch of concatenated lines that when you rotate one, it rotates all the following ones too). You can copy the [TextureBatcherTest](https://github.com/SilkCommunity/TrippyGL/tree/master/TrippyTests/TextureBatcherTest) project for this.

vpenades committed:
Exposed TextureBatcher Draw API using a Matrix3x2
Added drawing demo to TextureBatcherTest

vpenades commented:
Sorry for the late reply, wrapping the year at work it's being a nightmare.

> Wait what, what does it mean to multiply two 3x2 matrices, that's not math

Matrices in the System.Numerics namespace can be multiplied, and the practical effect is that the scale, rotation and translation of the parent matrix is "propagated" to the child matrix.  You can do the same without matrices, but it's way more verbose.

Also, using the System.Numerics matrices has the additional advantage that Microsoft is taking a lot of care into optimizing these structures, so in some cases, the generated code boils down to SIMD code, so although at first sight the code might look slower, it's actually much faster than working with position, scale and rotation separately.

I've added one more commit with the public methods for TextureBatcher, and added some code in TextureBatcherTest that showcases these draw calls.

vpenades commented:
So far, I think this is all for this PR

Having support for Matrix3x2 in TextureBatcher makes some other stuff I wanted to do much easier.

I have some working concepts on how to greatly simplify all the boiler plate code used to draw fonts, which uses the Matrix3x2 methods, but that's for some other time if you are willing to discuss about it.

ThomasMiz commented:
Look at 'im go 🤩🤩🤩
![look_at_it_go gif](https://github.com/SilkCommunity/TrippyGL/assets/32400648/d84c907f-deba-417f-8b91-bbc41570c037)
Ah crap, I broke the gif recorder.

This is the first code put into TrippyGL that wasn't written by me! Thank you!
